src/apps/bm_devkit/sdi12_example/user_code/exo3s_sensor.cpp

```diff
@@ -0,0 +1,256 @@
+#include "stm32_rtc.h"
+#include "task_priorities.h"
+#include "uptime.h"
+#include "usart.h"
+#include "util.h"
+#include "exo3s_sensor.h"
+#include "bristlefin.h"
+#include "stm32_io.h"
+
+SondeEXO3sSensor::SondeEXO3sSensor()
+    : d0_parser(4, "0D0!0"),
+      d1_parser(4, "0D1!0"),
+      d2_parser(2, "0D2!0"),
+      _latest_sample{} {
+}
+
+
+void SondeEXO3sSensor::init() {
+  printf("SDI-12 Initialization");
+  d0_parser.init();
+  d1_parser.init();
+  d2_parser.init();
+
+  PLUART::init(USER_TASK_PRIORITY);
+  PLUART::setEvenParity(); // data width is 7 bits + 1 parity bit = 8 bits
+  PLUART::enableDataInversion();
+  PLUART::setBaud(BAUD_RATE);
+  PLUART::setUseByteStreamBuffer(false);
+  PLUART::setUseLineBuffer(true);
+
+  PLUART::setTerminationCharacter(LINE_TERM);
+  // startTransactions - sdi12Tx (enable TX by setting OE LOW)
+  // endTransactions - sdi12Rx (disable TX by setting OE HIGH)
+  PLUART::enableTransactions(Bristlefin::sdi12Tx, Bristlefin::sdi12Rx);
+  PLUART::enable();
+}
+
+void SondeEXO3sSensor::sdi_wake(void) {
+  uint32_t timeStart;
+  timeStart = uptimeGetMs();
+  PLUART::disable();
+  //Set TX pin to output
+  PLUART::configTxPinOutput();
+  // HIGH at TX pin
+  PLUART::setTxPinOutputLevel();
+  // Wake - hold HIGH for 12 ms
+  timeStart = uptimeGetMs();
+  while(uptimeGetMs() - timeStart < breakTimeMin);
+  // Set TX pin back to TX (alternate) mode
+  PLUART::configTxPinAlternate();
+  // Re-enable UART
+  PLUART::enable();
+  printf("sdi_wake\n");
+}
+
+
+void SondeEXO3sSensor::sdi_break_mark(void) {
+  flush();
+  uint32_t timeStart;
+  PLUART::disable();
+  //Set TX pin to output
+  PLUART::configTxPinOutput();
+  // HIGH at TX pin
+  PLUART::setTxPinOutputLevel();
+  // Break - hold HIGH for 12 ms
+  timeStart = uptimeGetMs();
+  while(uptimeGetMs() - timeStart < breakTimeMin);
+  // LOW at TX pin
+  PLUART::resetTxPinOutputLevel();
+  // Mark - hold LOW for 9 ms
+  timeStart = uptimeGetMs();
+  while(uptimeGetMs() - timeStart < markTimeMin);
+  // Set TX pin back to TX (alternate) mode
+  PLUART::configTxPinAlternate();
+  // Re-enable UART
+  PLUART::enable();
+}
+
+
+void SondeEXO3sSensor::sdi_transmit(const char *ptr) {
+  // TX enable
+  PLUART::startTransaction();
+  sdi_break_mark();
+  PLUART::write((uint8_t *)ptr, strlen(ptr));
+  // TX disable
+  PLUART::endTransaction(100); // 50,28 ms to release/get some mutex/semaphore????
+}
+
+
+
+bool SondeEXO3sSensor::sdi_receive(void) {
+  uint64_t timeStart = uptimeGetMs();
+  memset(rxBuffer, 0, sizeof(rxBuffer));
+  while (!PLUART::lineAvailable() && (uptimeGetMs() - timeStart < receive_timeout)){
+    vTaskDelay(pdMS_TO_TICKS(5));
+  }
+  if (PLUART::lineAvailable()){
+    uint16_t len = PLUART::readLine(rxBuffer, sizeof(rxBuffer));
+    for (int i =0; i < len; i++){
+      rxBuffer[i] = rxBuffer[i] & 0x7F;
+    }
+    return true;
+  }
+  return false;
+}
+
+
+void SondeEXO3sSensor::sdi_cmd(int cmd) {
+  char result = sdiSuccess;
+
+  switch (cmd) {
```

Comment:
Can these cases have an enum associated with them? It is hard to understand what each number indicates,
something like:
```
enum {
  SDI_ADDRESS_INQUIRY_CMD,
  SDI_INDENTIFICATION_CMD,
  SDI_START_MEASUREMENT_CMD,
};
```

Reply:
I can see that! I'll make these an enum 👍